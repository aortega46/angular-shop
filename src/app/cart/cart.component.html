<button
  class="absolute right-2 top-2 z-50 flex h-12 w-12 cursor-pointer items-center justify-center rounded-full bg-blue-400 p-1"
  (click)="toggleCart()">
  <mat-icon class="material-icons-outlined" fontIcon="shopping_cart"></mat-icon>
</button>

<aside
  class=" bg-black p-8 top-0 right-0 transition {{
    isCartOpened
      ? 'translate-x-0 fixed h-full w-52'
      : 'translate-x-full fixed h-full'
  }}">
  <ul>
    @for (product of []; track $index) {
      <cart-item [product]="product" />
    }
  </ul>

  <button>
    <mat-icon
      class="material-icons-outlined"
      fontIcon="remove_shopping_cart"></mat-icon>
  </button>
</aside>
