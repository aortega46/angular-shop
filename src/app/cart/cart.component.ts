import { Component } from '@angular/core'
import { MatIconModule } from '@angular/material/icon'
import { CartItemComponent } from '../cart-item/cart-item.component'

@Component({
  selector: 'app-cart',
  standalone: true,
  imports: [MatIconModule, CartItemComponent],
  templateUrl: './cart.component.html',
})
export class CartComponent {
  isCartOpened: boolean = false

  toggleCart() {
    this.isCartOpened = !this.isCartOpened
  }
}
