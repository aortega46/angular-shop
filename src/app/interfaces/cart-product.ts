export interface CartProduct {
  id: number
  title: string
  price: number
  thumbnail: string
  quantity: number
}
