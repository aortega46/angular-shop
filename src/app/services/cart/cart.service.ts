import { Injectable } from '@angular/core'
import { Product } from '../../interfaces/product'
import { BehaviorSubject, Observable, of, switchMap } from 'rxjs'
import { CartProduct } from '../../interfaces/cart-product'

@Injectable({
  providedIn: 'root',
})
export class CartService {
  cart = new BehaviorSubject<CartProduct[]>([])

  addToCart(product: Product) {
    const currentCart = this.cart.getValue()
    const productInCartIndex = currentCart.findIndex(
      item => item.id === product.id
    )

    if (productInCartIndex >= 0) {
      const newCart = [
        ...currentCart.slice(0, productInCartIndex),
        {
          ...currentCart[productInCartIndex],
          quantity: currentCart[productInCartIndex].quantity + 1,
        },
        ...currentCart.slice(productInCartIndex + 1),
      ]

      return this.cart.next(newCart)
    }

    const newCart = [...currentCart, { ...product, quantity: 1 }]
    return this.cart.next(newCart)
  }

  checkProductInCart(id: number): Observable<boolean> {
    return this.cart.pipe(
      switchMap(cart => of(cart.some(item => item.id === id)))
    )
  }

  removeFromCart({ id }: { id: number }) {
    const currentCart = this.cart.getValue()
    const newCart = currentCart.filter(item => item.id !== id)
    this.cart.next(newCart)
  }

  clearCart() {
    this.cart.next([])
  }
}
