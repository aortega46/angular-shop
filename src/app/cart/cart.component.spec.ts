import { ComponentFixture, TestBed } from '@angular/core/testing'

import { CartComponent } from './cart.component'
import { CartService } from '../services/cart/cart.service'
import { CartProduct } from '../interfaces/cart-product'

describe('CartComponent', () => {
  let component: CartComponent
  let fixture: ComponentFixture<CartComponent>
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CartComponent],
    }).compileComponents()

    fixture = TestBed.createComponent(CartComponent)
    component = fixture.componentInstance
    fixture.detectChanges()

    compiled = fixture.nativeElement
  })

  it('should create', () => {
    expect(component).toBeTruthy()
  })

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('should toggleCart', () => {
    const intialState = component.isCartOpened

    component.toggleCart()
    expect(component.isCartOpened).not.toEqual(intialState)

    component.toggleCart()
    expect(component.isCartOpened).toEqual(intialState)
  })

  it('should call service when #addToCart', () => {
    const service = TestBed.inject(CartService)
    const spyAddToCart = jest.spyOn(service, 'addToCart')

    const mockProduct: CartProduct = {
      id: 1,
      title: 'Product 1',
      price: 10,
      thumbnail: 'test.png',
      quantity: 1,
    }

    component.addToCart(mockProduct)
    expect(spyAddToCart).toHaveBeenCalledWith(mockProduct)
  })

  it('should display cart items', () => {
    const cartItems = compiled.querySelectorAll('cart-item')

    expect(cartItems.length).toBe(1)
    expect(cartItems[0].textContent).toContain('Product 1')
    expect(cartItems[0].textContent).toContain('$10')
    expect(cartItems[0].querySelector('img')?.getAttribute('src')).toContain(
      'test.png'
    )
  })
})
