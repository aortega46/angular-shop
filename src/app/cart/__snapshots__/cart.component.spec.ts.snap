// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <button
    class="fixed right-4 top-2 z-50 flex h-12 w-12 cursor-pointer items-center justify-center rounded-full bg-blue-400 p-1"
  >
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons-outlined material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-name="shopping_cart"
      data-mat-icon-type="font"
      fonticon="shopping_cart"
      role="img"
    />
  </button>
  <aside
    class="bg-black fixed flex flex-col h-full items-center overflow-y-auto p-8 right-0 top-0 transition translate-x-full w-72"
  >
    <ul
      class="mb-2 mt-4 border-b-2 border-b-slate-800"
    >
      
    </ul>
    <button
      class="flex rounded bg-slate-800 px-4 py-2"
    >
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons-outlined material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-name="remove_shopping_cart"
        data-mat-icon-type="font"
        fonticon="remove_shopping_cart"
        role="img"
      />
    </button>
  </aside>
</div>
`;
